Add function `applyNM` to `Control.Monad.Util`.

jonathanknowles opened the pull request:
## Issue Number

ADP-1783

## Summary

This PR adds the `applyNM` function and property tests.

The `applyNM` function applies a monadic function **_n_** times to an initial value, returning the final result.

Its behaviour is similar to `applyN`, but applies a monadic function instead of a pure function, hence the addition of the `M` suffix:

```hs
applyN  :: forall   a.            Int -> (a ->   a) -> a ->   a
applyNM :: forall m a. Monad m => Int -> (a -> m a) -> a -> m a
```

jonathanknowles committed:
Add function `applyNM`.

This commit adds the `applyNM` function and property tests.

This function applies a monadic function @n@ times to an initial value,
returning the final result.

Its behaviour is similar to `applyN`, but applies a monadic function
instead of a pure function, hence the addition of the `M` suffix:

```
applyN  :: forall   a.            Int -> (a ->   a) -> a ->   a
applyNM :: forall m a. Monad m => Int -> (a -> m a) -> a -> m a
```

HeinrichApfelmus commented on lib/core/src/Control/Monad/Util.hs:
```diff
@@ -0,0 +1,37 @@
+-- ]
+-- @
+--
+applyNM :: forall m a. Monad m => Int -> (a -> m a) -> a -> m a
+applyNM n next = loop n
+  where
+    loop :: Int -> a -> m a
+    loop !i !a
+        | i <= 0 = pure a
+        | otherwise = loop (i - 1) =<< next a
```
```suggestion
applyNM :: forall m a. Monad m => Int -> (a -> m a) -> a -> m a
applyNM n next = (!! n) . iterate (next =<<) . pure
```

[Wholemeal programming](https://stackoverflow.com/questions/6957270/what-is-wholemeal-in-functional-programming) 🤓

This version is "obviously correct" — I think QuickChecking this one is a bit overkill. If you really want to, checking the fast version against the "obviously correct" one may be worth doing, though.

jonathanknowles replied:
> This version is "obviously correct"

It's definitely more concise! (Though not sure about the performance.)

There is one difference though with this more concise implementation: it fails with a run-time error if `n` is less than `0`. (Whereas the original function treats negative values in the same way as functions like `take`.)

>  If you really want to, checking the fast version against the "obviously correct" one may be worth doing, though.

Thanks, that's a nice idea! I've added a commit that does this:  bda9108e4c6db0a4545ffdd6d7de79a836b9f8c0.

HeinrichApfelmus reviewed (approved):
Looks good to me, but I find the implementation and QuickCheck a bit overkill.

jonathanknowles committed:
Add test that compares `applyNM` against a more concise implementation.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3477#discussion_r958381655

jonathanknowles committed:
Limit number of iterations for slow running tests.
